### PR TITLE
IV-97: Setup CI/CD pipeline

### PR DESCRIPTION
**Pull request by dversoza-ec**
we could improve the pipeline with parallel builds for multi-arch containers: https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners

### CODE REVIEW

File: .aws/ecs-task-definition-staging.json [hunk 1]
```diff
@@ -35,16 +44,14 @@
                     "valueFrom": "arn:aws:secretsmanager:us-east-1:363611146132:secret:/ec-private-gpt/staging-fjZ8Mt:QDRANT_URL::"
                 }
             ],
-            "ulimits": [],
             "logConfiguration": {
                 "logDriver": "awslogs",
                 "options": {
                     "awslogs-create-group": "true",
```
Comment (justinomora) [inline .aws/ecs-task-definition-staging.json:50]: @dversoza-ec : I saw the word `staging` being explicitly used in several places to access and/or define several resources like the `awslogs-group` or `ECS_CLUSTER`. This implies that we will have to create a similar `.json` file for the `production` environment. I was wondering what your thoughts are and/or if you think if would be a good idea to have a single file and replace any references to the environment name with an github actions environment secret? Thus, lines 51 (in this file) and line 14 (in `ecs-deploy-staging.yml`) could look like: 
- line 51: "awslogs-group": "/ecs/ec-private-gpt-$env.ENV_NAME-task-definition",
- line 14: ECS_CLUSTER: ec-private-gpt-${{ env.ENV_NAME }}

^I'm not fully sure if that's the correct syntax for inserting GH secrets, but the idea would be to have a single files for both environments. 

Would this suggestion make our task definition & deployment files less readable and harder to maintain?

Comment (dversoza-ec) [reply .aws/ecs-task-definition-staging.json:50]: We could use the same file for both and create input arguments (or environment variables) to manipulate the environment name. 

I'd only highlight that there are a few problems with that:

1. we must adjust every environment-mutable argument (i.e., memory, CPU, secrets, roles, etc.) 
2. we must always use `staging`/`production` in every resource name (i.e., networking resources and IAM roles) - which should be fine unless we have shared resources among them.
3. we must decide if we'll go full CI/CD with a single staging/production pipeline or if we'll have two pipelines - for separated pipelines, we would need two different workflow files (as production doesn't build, only deploys).

To make it faster, I'd suggest leaving the two environments separated. But if you think we should improve the code quality and reduce duplication (with a little extra time to adjust these settings), I could combine the deployments files!

Comment (dversoza-ec) [reply .aws/ecs-task-definition-staging.json:50]: Oh... one last thing... we would need a different renderer for the `task_definition.json` file. Currently, we aren't mutating it, except for the image tag. 

Terraform's `template_file()` function could do the work, but that will force us to use that sooner.

Comment (dversoza-ec) [reply .aws/ecs-task-definition-staging.json:50]: Maybe we could leave the task definition for later and work only on the deployment pipeline now ... If that's the case

Comment (justinomora) [reply .aws/ecs-task-definition-staging.json:50]: @dversoza-ec :  Cool! Thanks for the detailed response! Makes more sense to have them in separate files. Approving now!

**Review by justinomora — APPROVED** [review]
looks good!

**Comment by dversoza-ec** [comment]
Fixed health checks!! 

Screenshot 2024-02-29 at 3 46 46 PM

**Pull request merged**